ml/Database.cc

```diff
@@ -38,6 +38,80 @@ const char *ml::Database::SQL_SELECT_ANOMALY_EVENTS =
     "   after >= ?4 AND"
     "   before <= ?5;";
 
+const char *ml::Database::SQL_CREATE_ANOMALY_RATE_INFO_TABLE =
+    "CREATE TABLE IF NOT EXISTS anomaly_rate_info( "
+    "     host_id text NOT NULL, "
+    "     after int NOT NULL, "
+    "     before int NOT NULL, "
+    "     anomaly_rates text "
+    ");";
+
+const char *ml::Database::SQL_INSERT_BULK_ANOMALY_RATE_INFO =
+    "INSERT INTO anomaly_rate_info( "
+    "     host_id, after, before, anomaly_rates) "
+    " VALUES (?1, ?2, ?3, ?4);";
+   
+const char *ml::Database::SQL_SELECT_ANOMALY_RATE_INFO =
+    "SELECT  main.dimension_id, COALESCE("
+    " ((COALESCE(pre.avg,0) * (SELECT before-?2 FROM anomaly_rate_info WHERE after <= ?2 ORDER BY after DESC LIMIT 1)) + "
+    " (COALESCE(main.avg,0) * ((SELECT before FROM anomaly_rate_info WHERE before <= ?3 ORDER BY before DESC LIMIT 1) - "
+    "              (SELECT after FROM anomaly_rate_info WHERE after >= ?2 ORDER BY after ASC LIMIT 1))) + "
+    " (COALESCE(post.avg,0) * (SELECT ?3-after FROM anomaly_rate_info WHERE before >= ?3 OR "
+    "                ((SELECT before FROM anomaly_rate_info WHERE before <= ?3 ORDER BY before DESC LIMIT 1) "
+    "                 AND NOT EXISTS (SELECT before FROM anomaly_rate_info WHERE before >= ?3 ORDER BY before ASC LIMIT 1)) "
+    "              ORDER BY before ASC LIMIT 1))) / ( "
+    "  ((COALESCE(pre.avg,0)/(CASE WHEN COALESCE(pre.avg,1) = 0 THEN 0.0000001 ELSE COALESCE(pre.avg,1) END) * "
+    "   (SELECT before-?2 FROM anomaly_rate_info WHERE after <= ?2 ORDER BY after DESC LIMIT 1)) + "
```

Comment:
Hi Siamak!

What I think is missing here, is that if the average from the pre is 0, then we lose the range of it, which is needed in the final calculation.

The same goes for the post part.

In our simple case:

|after|before|a_dimension|
|------|---------|---------------|
|10|20|0|
|20|30|5|
|30|40|0|
|40|50|0|

If we request after = 15, before = 45, we need to include both the 15->20 range (5) in pre + the 40 -> 45 (5) from post even if the average in these ranges is 0 for the division part.

Please correct me if wrong, or I have understood it wrongly.

Reply:
I think you are right, as it came to my attention in the test break-down, too, and if we review some earlier commits, it was considering the real (shorter and proportional range) by having before-parameter and parameter-after instead of before-after. I changed it for a reason! but am putting it back to what it was and checking all the tests...

Reply:
Sounds all good after all tests passed, thanks for raising the issue.

Reply:
Seems to work much better, thanks!! Will continue to check!